DRAFT: enabling doctestplus

bsipocz opened the pull request:
This is a WIP, and I'm opening a PR only to enable @seberg to push to this branch directly.

bsipocz committed:
MAINT: making pytest with doctestplus collect without errors

bsipocz committed:
TMP: some fixes

bsipocz committed:
DOC: switching to use the plot directive

bsipocz committed:
MAINT: adding doctestplus to test requirements

seberg commented:
@bsipocz I pushed a small fix for a merge conflict.  But, could you remind me of the correct incantation (currently?), I already added:
```
 norecursedirs =
+    # Skip potential temporary environments:
+    build
+    benchmarks/env
```
to the `pytest.ini` (both can contain a conftest.py which fixed the below issue for it), but I keep getting import errors like:
```
_pytest.pathlib.ImportPathMismatchError: ('random', '/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/random.py', PosixPath('/Users/sebastianb/forks/numpy/numpy/typing/tests/data/pass/random.py'))
```
so I think I am probably just approaching running `pytest` wrong.

seberg committed:
DOC: Reset string function to default at end of test

seberg committed:
TST: Skip build and benchmark/envs folder for doctests

seberg committed:
DOC: Fix records doctests by improving docstrings

seberg commented:
There are now quite a few plots popping up again and failures due to deprecations.  Maybe we should just ignore deprecation warnings in the docs.  Or if we don't want to do that, just remove the examples for clearly deprecated functionality?

mattip commented:
Would a rebase help clear some of the collection issues?

seberg committed:
DOC: Add traceback start for doctestplus

seberg commented on numpy/conftest.py:
```diff
@@ -114,6 +114,10 @@ def check_fpu_mode(request):
+    # NumPy should probably use more `.. plot::` directives even in
+    # examples, but as long as we don't, do not block on `plt.show()`
+    import matplotlib
+    matplotlib.use("Agg")
```
@bsipocz where could this go and is this OK?  This fixes the `plt.show()` not popping up things.  There are some that can just be plot directives, but the histogram re-uses the data and has comments, etc. so I thought this is the easier path.

Maybe we need distinct runs, also so that we can make sure that `warnings` are not raised as errors?

bsipocz replied:
Indeed, this should be the fix for the mpl pop ups.

As for warnings, I think they really should be raised as errors, unless they are handled properly. For doctestplus we can use either `+IGNORE_WARNINGS` or `+SHOW_WARNINGS`

seberg commented:
I think the collection issues are maybe all expected, i.e. trying to find doctests in places we shouldn't look or just something to do with how I ran it.

bsipocz commented:
Thanks Sebastian for the commits! I'm sorry for being too slow with this, I'm trying to come back to it in a more timely manner.

seberg committed:
TST: Readd switching to matplotlib Agg backend when doctests are run

seberg committed:
DOC: Use doctest-skip for datasource examples which assume files exist

seberg committed:
DOC: Skip doctest for datasource

The file uses too many things that assume a layout and almost nothing
can be doctested anyway

seberg committed:
DOC: Add doctest-skip to internal/historic README

It just doesn't hurt, in case its run, this file is only interesting
historically and could probably be moved out as well.

seberg committed:
TST: Make extbuild doctest pass

seberg committed:
DOC: Add a doctest skip to a few lines that are not meant to be run

seberg committed:
DOC: Fix doctest for deprecated including show warnings directive

bsipocz committed:
DOC: fix linter

rgommers commented:
I'll note that the tool we currently use is `tools/refguide_check.py`. It is run in CircleCI only. It can also be run via `runtests.py` and should still be plugged into `spin test`.

If you want to iterate on CI, perhaps easiest to use that same job (or port it to a single GHA job).

mattip commented:
We now have `spin check-docs`. @bsipocz, @seberg should we close this?

bsipocz commented:
Yes, I think so. I think it's a pitty that we couldn't merge the approaches instead of having one tool doing the same another, less generically used was built up, but whatever.
